sdk/metric/reader.go

```diff
@@ -0,0 +1,92 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metric
+
+import (
+	"context"
+
+	"go.opentelemetry.io/otel/sdk/metric/export"
+	"go.opentelemetry.io/otel/sdk/metric/view"
+)
+
+// Reader is the interface used between the SDK and an
+// exporter.  Control flow is bi-directional through the
+// Reader, since the SDK initiates ForceFlush and Shutdown
+// while the initiates collection.  The Register() method here
+// informs the Reader that it can begin reading, signaling the
+// start of bi-directional control flow.
+//
+// Typically, push-based exporters that are periodic will
+// implement PeroidicExporter themselves and construct a
+// PeriodicReader to satisfy this interface.
+//
+// Pull-based exporters will typically implement Register
+// themselves, since they read on demand.
+type Reader interface {
+	// Register is called when the SDK is fully
+	// configured.  The Producer passed allows the
+	// Reader to begin collecting metrics using its
+	// Produce() method.
+	register(producer)
+
+	Collect(context.Context) (export.Metrics, error)
```

Comment:
I think this can be removed, it's not needed if the Register method, Producer type, and Produce method are exposed. When these are exposed, the ManualReader is only needed for simple applications; anything that wants manual reading behavior and also to control flush/shutdown has the option to manually implement the reader themselves, it's very simple.
 
In my branch, Collect() is an interface method that the (async/sync)x(cumulative/delta) instrument objects implement, used by the producer during Produce()

```
// Collector is an interface for things that produce Instrument data.
// One instrument may output more than one Instrument data by
// appending to `output`.
type Collector interface {
	Collect(sequence Sequence, output *[]Instrument)
}
```

In my branch I named what is `export` here as the package `sdk/metric/data`, so ^^^ is the `data.Collector` interface and `Collect()` collects is how one API-level instrument outputs multiple View-level metric data objects. The user would not use this, it's between the `*meter` and the internal packages.

Reply:
`Collect` is a required method for a reader: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#collect